design/baremetal-operator/unmanaged-annotation.md

```diff
@@ -0,0 +1,106 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Add support for Unmanaged annotation
+
+## Status
+
+implementable
+
+## Summary
+
+Provide a way to prevent management of BaremetalHost resources
+after provisioning is completed, to facilitate the pivot of
+BaremetalHost resources in a multi-tier deployment.
+
+## Motivation
+
+In a multi-tier deployment where one cluster deploys another, the "parent"
+cluster will contain BMH resources for initial provisioning,
+but the "child" cluster will later contain BMH resources that reference the
+same physical hosts.
+
+In this scenario it's necessary to prevent management operations from the
+parent cluster such as asserting power state, or BMH actions on the child
+cluster such as the [reboot annotation](reboot-interface.md) may fail due
+to unwanted BMC interactions from the parent cluster.
+
+There is an existing
+[pause annotation](https://github.com/metal3-io/baremetal-operator/blob/master/docs/api.md#pausing-reconciliation)
+which pauses reconciliation of the BMH resources, but this does not remove
+the underlying Ironic host, so power management actions may be performed
+even when the BMH is marked as `paused`.
+
+### Goals
+
+* Add an API to disable management of a BMH on the parent cluster, including
+  all power management actions
+* Make it possible to delete a BMH resource in this "unmanaged" without
+  triggering deprovisioning
+* Ensure it is possible to restart management of BMH resources (in the case
+  where they are not deleted from the parent cluster)
+* Avoid changing behavior of the existing `paused` annotation since
+
+### Non-Goals
+
+* Any coordination between the tiers of BMH resources, that has to be handled externally
+
+## Proposal
+
+### Expected workflow
+
+* User creates BMH resource(s) in parent cluster
+* Provisioning of BMH triggered, which results in a running child cluster
+* Parent cluster BMH resources marked as unmanaged
+* Child cluster BMH resources created, with BMC credentials, but marked
+  externallyProvisioned: true
+
+At this point, the physical hosts are owned by the child cluster BMH, but the
+inventory still exists in the parent cluster.
+
+In the event that all of the child cluster hosts require reprovisioning, it
+would be necessary to remove the maintenance/unmanaged flag, or
+delete/re-create the BMH resource so that the process starts again at the
+beginning.
+
+### API changes
+
+Add support for a new annotation, where the key is `baremetalhost.metal3.io/unmanaged`
```

Comment:
A requirement I thought I heard on the call the other day (@maelk correct me if this is wrong) is that for pivoting we need a way of knowing when the host has become unmanaged, so that the paused annotation can then be added as well afterwards. I'm unsure if this is true though, because I would have assumed the pause was mainly for the Host in the target cluster and the unmanage is for the Host in the bootstrap cluster.

If this is a requirement then we will need to figure out how to reflect progress in the Host status. If that ends up being a separate field then it probably makes sense to add this to the Spec rather than an annotation. If not then I prefer keeping it as an annotation.

Reply:
Thanks @zaneb I've also been thinking about this - and the fact that we already have a BMH [Unmanaged State](https://github.com/metal3-io/baremetal-operator/blob/master/apis/metal3.io/v1alpha1/baremetalhost_types.go#L160)

This is kind of a weird non-state though, because we don't expose this to the user at all via `status.provisioning.state`

It's also not possible to transition from any state other than [Created](https://github.com/metal3-io/baremetal-operator/blob/master/docs/BaremetalHost_ProvisioningState.png) into this state (if you remove BMC creds from an existing BMH in `Provisioned` state, we go into `registration error`)

When discussing this with @dhellmann I think we said it would be best to avoid adding any transition from `$state->Unmanaged`, but I'm wondering if we should revisit that idea?

If we modified the `Unmanaged` behavior so that:

- We move to `Unmanaged` state whenever there are no BMC details, or when the `unmanaged` annotation is set
- When moving to `Unmanaged` state we remove any corresponding host in Ironic
- Set `status.provisioning.state` to reflect when we're in `Unmanaged` state (either on create, BMC details are removed, or when the annotation is added)
- When moving from the Unmanaged State, always go to `Registering`, except when an existing Image exists in the status - then we'd add a new transition to `Provisioned`

WDYT?

Reply:
> Thanks @zaneb I've also been thinking about this - and the fact that we already have a BMH [Unmanaged State](https://github.com/metal3-io/baremetal-operator/blob/master/apis/metal3.io/v1alpha1/baremetalhost_types.go#L160)
> 
> This is kind of a weird non-state though, because we don't expose this to the user at all via `status.provisioning.state`

Yes we do: https://github.com/metal3-io/baremetal-operator/blob/master/controllers/metal3.io/host_state_machine.go#L275
Proof: https://bugzilla.redhat.com/show_bug.cgi?id=1936966#c0

> It's also not possible to transition from any state other than [Created](https://github.com/metal3-io/baremetal-operator/blob/master/docs/BaremetalHost_ProvisioningState.png) into this state (if you remove BMC creds from an existing BMH in `Provisioned` state, we go into `registration error`)

We stay in Provisioned and the operational status goes to error with error type "registration error". (Being pedantic because prior to metal3-io/baremetal-operator#388 there was a RegistrationError provisioning state.)

> When discussing this with @dhellmann I think we said it would be best to avoid adding any transition from `$state->Unmanaged`, but I'm wondering if we should revisit that idea?
> 
> If we modified the `Unmanaged` behavior so that:
> 
> * We move to `Unmanaged` state whenever there are no BMC details, or when the `unmanaged` annotation is set
> * When moving to `Unmanaged` state we remove any corresponding host in Ironic
>  * Set `status.provisioning.state` to reflect when we're in `Unmanaged` state (either on create, BMC details are removed, or when the annotation is added)
> * When moving from the Unmanaged State, always go to `Registering`, except when an existing Image exists in the status - then we'd add a new transition to `Provisioned`
> 
> WDYT?

The main problem with changing the provisioning state is that you then have to figure out which state you came from in order to go back. Basically we'd have to resurrect [this abomination](https://github.com/metal3-io/baremetal-operator/pull/388/commits/0e0d2af24c901b2933b8ff0e2d4f1d97f67e2056#diff-eeed3703d04e4c23a7d7af8cd0b7931b6b7990f23d826c49bdbc31c5f0a50291L214-L230). Limiting it to only returning to either Registration or Provisioned definitely makes it less bad than it could be. Whether the host has been provisioned can currently be determined reliably by looking at the existence of an image in the status (although I feel that this may be a mistake, and we possibly should put it in the status as soon as we start provisioning).

I guess the other concern would be for the user does it make sense that they could have got into this state (and therefore have to get out of it) by either of two completely different methods. Currently the OperationalStatus in the Unmanaged state is `discovered`; we probably want to change that to `inactive` or something for this case. Maybe that would be enough to distinguish them, although we did recently hide the operational status from the default `kubectl get` output.

Another option would be to change *only* the operational status (and leave the provisioning state where it was).

I'm certainly in favour of not having to add yet another status field.

Reply:
> Yes we do: https://github.com/metal3-io/baremetal-operator/blob/master/controllers/metal3.io/host_state_machine.go#L275

Thanks @zaneb - turns out that only works when `externallyProvisioned: true` is set, which is probably a bug - if you create a BMH without that set (and without any BMC details), the Operational State ends up empty.

Reply:
So yeah, this is a bug because we try to [create the provisioner](https://github.com/metal3-io/baremetal-operator/blob/master/controllers/metal3.io/baremetalhost_controller.go#L216) even when there are no BMC details - this results in an error, and we never get as far as setting any status.

I'm not yet sure why adding `externallyProvisioned` works around that though.

Reply:
It's because of #748.

Reply:
Resolved via https://github.com/metal3-io/baremetal-operator/pull/824